File: code/416.partition-equal-subset-sum.java
/*
 * @lc app=leetcode id=416 lang=java
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
class Solution {
    public boolean canPartition(int[] nums) {
        if (nums.length == 0 || nums == null) {
            return false;
        }
        int n = nums.length;
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        int space = sum / 2;
        int[] dp = new int[space + 1];
        for (int i = 0; i < n; i++) {
            for (int j = space; j >= nums[i]; j--) {
                dp[j] = Math.max(dp[j], dp[j - nums[i]] + nums[i]);
            }

            if (dp[space] == space) {
                return true;
            }
        }
        return false;
    }
}
// @lc code=end

File: code/416.partition-equal-subset-sum.go
/*
 * @lc app=leetcode id=416 lang=golang
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
func canPartition(nums []int) bool {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	if sum%2 == 1 {
		return false
	}

	space := sum / 2
	dp := make([]int, space+1)
	for _, num := range nums {
		for j := space; j >= num; j-- {
			dp[j] = max(dp[j], dp[j-num]+num)
		}

		if dp[space] == space {
			return true
		}
	}
	return false
}

// @lc code=end

